Redesign Lumo variants

jouni opened the pull request:
Base the size of the handle in sizing properties instead of space properties.

Redesigned “small” variant – now just 1px divider and the handle is only visible on hover. The handle is now always the same size for normal and small, only the splitter part changes.

Rename “subtle” to “minimal” (align with other element variant names). The minimal variant doesn’t have any visible splitter, and the handle is only visible on hover.

Update demos to reflect the changes.

This is a breaking change, so I wanted to discuss this with a pull request before merging. The motivation for doing this change came from the “kitchen sink” test app that I built in the vaadin-lumo-styles repo, and these styles made the most sense for a clean application layout.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/vaadin-split-layout/100)
<!-- Reviewable:end -->

web-padawan commented:
What goals do we pursue by hiding the handle?

The Lumo theme variants demo itself looks now as a couple of geometric shapes and my initial feeling once I opened it was "something looks definitely broken here, do I need to install dependencies etc".

Furthermore, depending on hover makes me sad regarding mobile devices support. 1px gray border does not seem to be something encouraging user to interact with, it is just barely noticeable after all. 

I tried this with my android device and I was able to reveal handle on tap but rather occasionally. I clicked the left bar in the "Small" section and managed to do it, when I started clicking on the right one and the handle did not appear, but instead the handle for the left bar persists.

Overall UX impression is rather confusing to me, and the change is probably worth a couple of more UX tests.

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-split-layout/100#-:-L5h9M6VNzSZHm1Q-lN_:b54hmj)*
<!-- Sent from Reviewable.io -->

platosha commented:
Agree with @web-padawan. The handle is critical UX feature of this element, hiding it makes the split-layout appear as non-resizable. As mentioned, reveal on hover is unfriendly for touch input devices.

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[demo/split-layout-lumo-theme-demos.html, line 19 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-split-layout/100#-L5hFPk9yHb9IHvnhbKo:-L5hFPkAAZ_RMsQAjybl:bb4e1jr) ([raw file](https://github.com/vaadin/vaadin-split-layout/blob/833b46e4c5110d47f34a3c21c787288b97c7a9f4/demo/split-layout-lumo-theme-demos.html#L19)):*
> ```HTML
> 
>           div:first-child {
>             background-color: var(--lumo-shade-5pct);
> ```

In the demos, style adjustments on top of default look could be confusing. Users might expect this background color in their layouts too.

Please avoid if possible. Maybe use different content in the panels instead, as seen in the other demos?

---

*[demo/split-layout-lumo-theme-demos.html, line 47 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-split-layout/100#-L5hGWr7bNBOmdqR9W30:-L5hGWr7bNBOmdqR9W31:brhumzt) ([raw file](https://github.com/vaadin/vaadin-split-layout/blob/833b46e4c5110d47f34a3c21c787288b97c7a9f4/demo/split-layout-lumo-theme-demos.html#L47)):*
> ```HTML
> 
>           div:first-child {
>             background-color: var(--lumo-shade-5pct);
> ```

Same here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-split-layout/100#-:-L5hDXImhlQIuQuCLPN_:b-80aqkx)*
<!-- Sent from Reviewable.io -->

jouni commented:
> Furthermore, depending on hover makes me sad regarding mobile devices support. 1px gray border does not seem to be something encouraging user to interact with, it is just barely noticeable after all.

Do consider that these are variants, not the default style. The responsibility for using these rely on the developer/designer working on any given app/layout. Touch devices always require different considerations than other pointer devices. But the default style is IMO a good compromise to support both sufficiently.

> What goals do we pursue by hiding the handle?

> The handle is critical UX feature of this element, hiding it makes the split-layout appear as non-resizable. As mentioned, reveal on hover is unfriendly for touch input devices.

Having an “invisible” split region is very common in many apps. MacOS Finder, Google Drive, browser developer tools, just to name a few. Many business apps tend to have these, as there is a need to try to maximize the number of usable pixels on the screen and to minimize unnecessary visual noise.

So the UX is already established on the desktop. For touch devices, there are less common examples available, and the ones that I know of look pretty much like our default variation.

Usually, there isn’t a visible handle even when hovering, but I thought it would be a nice improvement if someone would end up using those on a touch device, so they would get a hint that yes, you can actually drag this (if they tap on the divider).

--

You can argue we should not have these built-in if you think it should be left for the app developer to handle. But IMO, there are small details to consider when implementing a 1px/0px splitter, so it’s better if we have those already implemented when someone needs to use that.

--

I think you are raising an important topic, which we have discussed recently with Marc, that we should eventually provide design guidelines for all of our UI components so that designers and devs know when some variant is suitable to use and what caveats it possibly has. It’s part of the ongoing design systems/productization opportunity exploration we are doing.

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[demo/split-layout-lumo-theme-demos.html, line 19 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-split-layout/100#-L5hFPk9yHb9IHvnhbKo:-L5hJX5vLHKR2H0xeIoZ:b-474wth) ([raw file](https://github.com/vaadin/vaadin-split-layout/blob/833b46e4c5110d47f34a3c21c787288b97c7a9f4/demo/split-layout-lumo-theme-demos.html#L19)):*
> The Lumo theme variants demo itself looks now as a couple of geometric shapes and my initial feeling once I opened it was "something looks definitely broken here, do I need to install dependencies etc".

Yeah, true. Any suggestions for the content? Red bg instead? :stuck_out_tongue_closed_eyes: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-split-layout/100#-:-L5hKNRDg6KI5CewnaMy:bbu4pkw)*
<!-- Sent from Reviewable.io -->

platosha commented:
OK fine by me then.

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[demo/split-layout-lumo-theme-demos.html, line 19 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-split-layout/100#-L5hFPk9yHb9IHvnhbKo:-L5hbe67SbEYJ1Hrv_C-:bndy7f3) ([raw file](https://github.com/vaadin/vaadin-split-layout/blob/833b46e4c5110d47f34a3c21c787288b97c7a9f4/demo/split-layout-lumo-theme-demos.html#L19)):*
<details><summary><i>Previously, jouni (Jouni Koivuviita) wrote…</i></summary><blockquote>

> The Lumo theme variants demo itself looks now as a couple of geometric shapes and my initial feeling once I opened it was "something looks definitely broken here, do I need to install dependencies etc".

Yeah, true. Any suggestions for the content? Red bg instead? :stuck_out_tongue_closed_eyes: 
</blockquote></details>

Maybe something like:

```html
<div>First</div>
<div>Second</div>
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-split-layout/100#-:-L5hbQtlCpy8k2ln9j-r:b8etatn)*
<!-- Sent from Reviewable.io -->

platosha commented:
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[theme/lumo/vaadin-split-layout.html, line 23 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-split-layout/100#-L5hcivgajTer2j8j9iQ:-L5hcivgajTer2j8j9iR:b-24ht8) ([raw file](https://github.com/vaadin/vaadin-split-layout/blob/833b46e4c5110d47f34a3c21c787288b97c7a9f4/theme/lumo/vaadin-split-layout.html#L23)):*
> ```HTML
>       [part="handle"]::after {
>         content: "";
>         position: absolute;
> ```

Any reason for using `absolute` here? Can we continue with flexbox instead?

Also, these `45%` coord values below look very suspicious. Something like `width: 10%; or `flex-basis: 10%;` would be more readable IMO.

---

*[theme/lumo/vaadin-split-layout.html, line 61 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-split-layout/100#-L5hdwpYnPlP_qrw-3Xm:-L5hdwpYnPlP_qrw-3Xn:b-p977en) ([raw file](https://github.com/vaadin/vaadin-split-layout/blob/833b46e4c5110d47f34a3c21c787288b97c7a9f4/theme/lumo/vaadin-split-layout.html#L61)):*
> ```HTML
>       /* Small/minimal */
> 
>       :host([theme~="small"]) > [part="splitter"] {
> ```

DOM level of parts may be subject to change in future. Could you avoid using `>` combinator here and below?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-split-layout/100)*
<!-- Sent from Reviewable.io -->

platosha commented:
Review status: 0 of 2 files reviewed at latest revision, 5 unresolved discussions.

---

*[theme/lumo/vaadin-split-layout.html, line 30 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-split-layout/100#-L5hf9SEKbZDhLD7mYAu:-L5hf9SEKbZDhLD7mYAv:b-c6x2vv) ([raw file](https://github.com/vaadin/vaadin-split-layout/blob/833b46e4c5110d47f34a3c21c787288b97c7a9f4/theme/lumo/vaadin-split-layout.html#L30)):*
> ```HTML
>         border-radius: var(--lumo-border-radius);
>         background-color: var(--lumo-contrast-30pct);
>         transition: 0.1s all;
> ```

BTW transitions for `all` may lead to unexpected results. Would rather keep that assigned to known properties, if possible.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-split-layout/100)*
<!-- Sent from Reviewable.io -->

web-padawan commented:
OK now I can see the point, but let's at least add some descriptions here in the demos, until we can refer to the documentation you mentioned.

One note about the naming: the "minimal" looks good but the "small" could be probably changed to something more descriptive.

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[demo/split-layout-lumo-theme-demos.html, line 10 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-split-layout/100#-L5hfNByjnoOIQKS4-QJ:-L5hfNByjnoOIQKS4-QK:b-f28aay) ([raw file](https://github.com/vaadin/vaadin-split-layout/blob/833b46e4c5110d47f34a3c21c787288b97c7a9f4/demo/split-layout-lumo-theme-demos.html#L10)):*
> ```HTML
> 
>     <h3>Small</h3>
>     <vaadin-demo-snippet id='split-layout-lumo-theme-demos-small'>
> ```

nit: please also change from single to double quotes

---

*[demo/split-layout-lumo-theme-demos.html, line 38 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-split-layout/100#-L5hfWtvK8CFmal9xz7H:-L5hfWtvK8CFmal9xz7I:b-f28aay) ([raw file](https://github.com/vaadin/vaadin-split-layout/blob/833b46e4c5110d47f34a3c21c787288b97c7a9f4/demo/split-layout-lumo-theme-demos.html#L38)):*
> ```HTML
> 
>     <h3>Minimal</h3>
>     <vaadin-demo-snippet id='split-layout-lumo-theme-demos-minimal'>
> ```

nit: please also change from single to double quotes

---

*[theme/lumo/vaadin-split-layout.html, line 64 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-split-layout/100#-L5hglgBObrUIb2nQPkf:-L5hglgBObrUIb2nQPkg:bv4ojva) ([raw file](https://github.com/vaadin/vaadin-split-layout/blob/833b46e4c5110d47f34a3c21c787288b97c7a9f4/theme/lumo/vaadin-split-layout.html#L64)):*
> ```HTML
>         min-width: 1px;
>         min-height: 1px;
>         background-color: var(--lumo-contrast-10pct);
> ```

consider using `border` instead. Using the `background` is risky, for example in desktop Safari 11, I pressed "zoom out" twice and the line disappeared.

FYI, the zoom issue with the background used like this previously existed in for `paper-input` underline, so we should ensure we do not use it in Material theme.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-split-layout/100#-:-L5hfGm5LtHxhfmmjHHP:b-i6g0zl)*
<!-- Sent from Reviewable.io -->

web-padawan commented:
The handle styles broken in Firefox:

![Screen Shot 2018-02-19 at 12.53.36.png](https://files.reviewableusercontent.io/images/reviewable/10589913/oyf4wJoyODg0dpAnQE0FI2_f/Screen%20Shot%202018-02-19%20at%2012.53.36.png)

---

Review status: all files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-split-layout/100#-:-L5hiIueyMJJg0M9AIe_:bfv1uh2)*
<!-- Sent from Reviewable.io -->

jouni commented:
> One note about the naming: the "minimal" looks good but the "small" could be probably changed to something more descriptive.

I can’t think of a good name. I could swap ”minimal” for the one with the visible line, and use “invisible” for the other. Not sure if that’s any better. Kept the names as they are for now.

> The handle styles broken in Firefox

Can’t reproduce with Firefox 58.0.2, works as expected.

---

Review status: 0 of 4 files reviewed at latest revision, 8 unresolved discussions.

---

*[demo/split-layout-lumo-theme-demos.html, line 10 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-split-layout/100#-L5hfNByjnoOIQKS4-QJ:-L5wwTzUSxPIdbc4soux:bm13htz) ([raw file](https://github.com/vaadin/vaadin-split-layout/blob/833b46e4c5110d47f34a3c21c787288b97c7a9f4/demo/split-layout-lumo-theme-demos.html#L10)):*
<details><summary><i>Previously, web-padawan (Sergey Kulikov) wrote…</i></summary><blockquote>

nit: please also change from single to double quotes
</blockquote></details>

Done, for all the demo pages.

---

*[demo/split-layout-lumo-theme-demos.html, line 19 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-split-layout/100#-L5hFPk9yHb9IHvnhbKo:-L5wwX0w59jAx1pUj16y:b-896fix) ([raw file](https://github.com/vaadin/vaadin-split-layout/blob/833b46e4c5110d47f34a3c21c787288b97c7a9f4/demo/split-layout-lumo-theme-demos.html#L19)):*
<details><summary><i>Previously, platosha (Anton Platonov) wrote…</i></summary><blockquote>

Maybe something like:

```html
<div>First</div>
<div>Second</div>
```
</blockquote></details>

Done.

---

*[demo/split-layout-lumo-theme-demos.html, line 38 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-split-layout/100#-L5hfWtvK8CFmal9xz7H:-L5wwZUme7lE4rpjCcsx:b-896fix) ([raw file](https://github.com/vaadin/vaadin-split-layout/blob/833b46e4c5110d47f34a3c21c787288b97c7a9f4/demo/split-layout-lumo-theme-demos.html#L38)):*
<details><summary><i>Previously, web-padawan (Sergey Kulikov) wrote…</i></summary><blockquote>

nit: please also change from single to double quotes
</blockquote></details>

Done.

---

*[demo/split-layout-lumo-theme-demos.html, line 47 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-split-layout/100#-L5hGWr7bNBOmdqR9W30:-L5ww_h71n3GEAPz2wMc:b-896fix) ([raw file](https://github.com/vaadin/vaadin-split-layout/blob/833b46e4c5110d47f34a3c21c787288b97c7a9f4/demo/split-layout-lumo-theme-demos.html#L47)):*
<details><summary><i>Previously, platosha (Anton Platonov) wrote…</i></summary><blockquote>

Same here.
</blockquote></details>

Done.

---

*[theme/lumo/vaadin-split-layout.html, line 23 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-split-layout/100#-L5hcivgajTer2j8j9iQ:-L5wwb2J2yHo7zN44sGK:b-wb59c2) ([raw file](https://github.com/vaadin/vaadin-split-layout/blob/833b46e4c5110d47f34a3c21c787288b97c7a9f4/theme/lumo/vaadin-split-layout.html#L23)):*
<details><summary><i>Previously, platosha (Anton Platonov) wrote…</i></summary><blockquote>

Any reason for using `absolute` here? Can we continue with flexbox instead?

Also, these `45%` coord values below look very suspicious. Something like `width: 10%; or `flex-basis: 10%;` would be more readable IMO.
</blockquote></details>

I was trying to make sure the visible handle element renders evenly on both sides of the split, by using offsets from both directions of the parent element. But I don’t think it had any noticeable effect, so I reverted back to flex container ​like you suggested. Also fixed the other dimension now to 4px. Let’s see if that becomes a problem, we can always open that up to be affected by Lumo sizing properties later on.

---

*[theme/lumo/vaadin-split-layout.html, line 30 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-split-layout/100#-L5hf9SEKbZDhLD7mYAu:-L5wx3LzDq25RAqnlx-6:b-896fix) ([raw file](https://github.com/vaadin/vaadin-split-layout/blob/833b46e4c5110d47f34a3c21c787288b97c7a9f4/theme/lumo/vaadin-split-layout.html#L30)):*
<details><summary><i>Previously, platosha (Anton Platonov) wrote…</i></summary><blockquote>

BTW transitions for `all` may lead to unexpected results. Would rather keep that assigned to known properties, if possible.
</blockquote></details>

Done.

---

*[theme/lumo/vaadin-split-layout.html, line 61 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-split-layout/100#-L5hdwpYnPlP_qrw-3Xm:-L5woxVxXNuJxuW4qVE9:b-mv6nnh) ([raw file](https://github.com/vaadin/vaadin-split-layout/blob/833b46e4c5110d47f34a3c21c787288b97c7a9f4/theme/lumo/vaadin-split-layout.html#L61)):*
<details><summary><i>Previously, platosha (Anton Platonov) wrote…</i></summary><blockquote>

DOM level of parts may be subject to change in future. Could you avoid using `>` combinator here and below?
</blockquote></details>

Without the combinator, nested split layouts will inherit the theme from their parent in polyfilled browsers.

---

*[theme/lumo/vaadin-split-layout.html, line 64 at r1](https://reviewable.io:443/reviews/vaadin/vaadin-split-layout/100#-L5hglgBObrUIb2nQPkf:-L5wx54azbaUnQCkjBJT:b-896fix) ([raw file](https://github.com/vaadin/vaadin-split-layout/blob/833b46e4c5110d47f34a3c21c787288b97c7a9f4/theme/lumo/vaadin-split-layout.html#L64)):*
<details><summary><i>Previously, web-padawan (Sergey Kulikov) wrote…</i></summary><blockquote>

consider using `border` instead. Using the `background` is risky, for example in desktop Safari 11, I pressed "zoom out" twice and the line disappeared.

FYI, the zoom issue with the background used like this previously existed in for `paper-input` underline, so we should ensure we do not use it in Material theme.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-split-layout/100#-:-L5wjZcLX2_8QmzMnbeF:b-ii75jl)*
<!-- Sent from Reviewable.io -->

platosha commented:
Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-split-layout/100)*
<!-- Sent from Reviewable.io -->

web-padawan commented:
OK now looks good in Firefox.

---

Reviewed 4 of 4 files at r2.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-split-layout/100#-:-L5xlOoHl39mFhognnQb:bsmresi)*
<!-- Sent from Reviewable.io -->

web-padawan commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-split-layout/100#-:-L5xm3NY3Qsor6lzAAMX:bnfp4nl)*
<!-- Sent from Reviewable.io -->

platosha commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-split-layout/100#-:-L5xnN5sXecQhNFzc6bK:bnfp4nl)*
<!-- Sent from Reviewable.io -->
